nixos/modules/services/web-apps/nextcloud.nix

```diff
@@ -309,10 +316,31 @@ in {
         }
       ];
 
-      warnings = optional (cfg.poolConfig != null) ''
-        Using config.services.nextcloud.poolConfig is deprecated and will become unsupported in a future release.
-        Please migrate your configuration to config.services.nextcloud.poolSettings.
-      '';
+      warnings = []
+        ++ (optional (cfg.poolConfig != null) ''
+          Using config.services.nextcloud.poolConfig is deprecated and will become unsupported in a future release.
+          Please migrate your configuration to config.services.nextcloud.poolSettings.
+        '')
+        ++ (optional (versionOlder cfg.package.version "18") ''
+          You're currently deploying an older version of Nextcloud. This may be needed
+          since Nextcloud doesn't allow major version upgrades across multiple versions (i.e. an
+          upgrade from 16 is possible to 17, but not to 18).
+
+          Please deploy this to your server and wait until the migration is finished. After
+          that you can deploy to the latest Nextcloud version available.
+        '');
+
+      services.nextcloud.package = with pkgs;
+        mkDefault (
+          if pkgs ? nextcloud
```

Comment:
`pkgs ? nextcloud` will always be false because it's removed.  Should this be `cfg.package`?

Reply:
It will be false unless it's defined in an overlay. If that's the case, an assertion will be thrown. This is done to make sure that no downgrade happens (e.g. Nextcloud 18.x in the overlay, but Nextcloud 17.x selected for `nextcloud.package`). If one wants to ues `pkgs.nextcloud` (from an overlay), `services.nextcloud.package` should be set to `pkgs.nextcloud` explicitly.